### PR TITLE
Support generated-only symbols in SymbolFinder

### PR DESCRIPTION
**Pull request by 333fred**
SymbolFinder was bailing out if there were no syntactic matches to the requested filter, avoiding creating a compilation at all if no document contained anything that could be named that symbol. Unfortunately, this optimization does not work if a project has source generators, as the generated documents are not checked.

Fixes https://github.com/dotnet/roslyn/issues/63375.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-ide for review.

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs [hunk 1]
```diff
@@ -56,17 +56,22 @@ internal static partial class DeclarationFinder
...
-                // overload that quickly uses the direct bloom-filter identifier maps in the index.  If it's nto an 
-                // exact name search, then we will run the query's predicate over every DeclaredSymbolInfo stored in
-                // the doc.
-                var containsSymbol = isExactNameSearch
-                    ? await project.ContainsSymbolsWithNameAsync(query.Name!, cancellationToken).ConfigureAwait(false)
-                    : await project.ContainsSymbolsWithNameAsync(query.GetPredicate(), filter, cancellationToken).ConfigureAwait(false);
-
-                if (!containsSymbol)
-                    return;
+                // This shortcut only works if there cannot be any generated documents. If there are generated documents, then
```
Comment (333fred) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs:59]: Recommend reviewing with whitespace off. Diff is much cleaner.

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs [hunk 2]
```diff
@@ -56,17 +56,22 @@ internal static partial class DeclarationFinder
...
+                // we'll need to realize the compilation to get them before concluding the symbol doesn't exist.
+                if (!project.State.SourceGenerators.Any())
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs:61]: this change cannot be made.  Practically all .net6 projects have source generators, so this will cause us to have a perf hit there no matter what.  We discovered just how bad this was with navigate to, and are still spending time trying to claw the performance back.  

My position is that we can only add SG support when we can do so in a manner that comes with no perf impact (or minimal perf impact) in the case that people *have* source generator assemblies, but aren't actually generating anything.  

In other words, if i'm on .net6/.net7 and i'm not taking advantage of SGs (but their assemblies are still connected to my project) i should take no (or minimal) perf hit.

This code will not do that.  We lose a MAJOR performance benefit here because we can no longer do this extremely quick syntactic check (which is the entire basis of FAR performance).

**Review by CyrusNajmabadi — CHANGES_REQUESTED** [review]
Hard block on this for now.  It's better to have incomplete results (which we could at least notify the user of) then lose hte ability to use syntactic indices to avoid unnecessarily creating compilations (hugely hugely hugely expensive).

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations_CustomQueries.cs [hunk 3]
```diff
@@ -100,7 +101,7 @@ public static async Task<IEnumerable<ISymbol>> FindSourceDeclarationsAsync(Proje
 
             using (Logger.LogBlock(FunctionId.SymbolFinder_Project_Predicate_FindSourceDeclarationsAsync, cancellationToken))
             {
-                if (await project.ContainsSymbolsWithNameAsync(query.GetPredicate(), filter, cancellationToken).ConfigureAwait(false))
+                if (project.State.SourceGenerators.Any() || await project.ContainsSymbolsWithNameAsync(query.GetPredicate(), filter, cancellationToken).ConfigureAwait(false))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations_CustomQueries.cs:104]: effectively, this is the same as pushing the boolean check into ContainsSymbolsWithNameAsync and basically having that function go "whelp... if htere are source generators, guess i can't do anything smart/fast".  We need to *not* go down that path.  This may require some sort of smarts on behalf of SGs to make it possible to have efficient features along with SGs. But we can't lose efficiency just when an SG is present.

Tagging @jasonmalinowski as this keeps coming up and is very concerning to me :(

**Comment by 333fred** [comment]
While the right approach for SymbolFinder is considered, I've submitted https://github.com/dotnet/roslyn/pull/63386/ so that O# can work around the issue.

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs:61]: So it's not clear to me what this API is used for precisely -- is it intended as a public form of a Navigate To API, or is this underneath some other feature we already have? I agree with @CyrusNajmabadi we need to be more careful here, or at least really convince ourselves we can't do something smarter.

I don't see any internal callers into Roslyn itself, but I'm doing that via a grep search because of all the overloads. So maybe I'm missing something?

Comment (333fred) [reply src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs:61]: >  is it intended as a public form of a Navigate To API

Yes, this is what OmniSharp uses it for.
